framework/src/transfers/MultiAppCopyTransfer.C

```diff
@@ -51,204 +52,99 @@ MultiAppCopyTransfer::initialSetup()
-
-        // Only works with a serialized mesh to transfer from!
-        mooseAssert(from_sys.get_mesh().is_serial(), "MultiAppCopyTransfer only works with ReplicatedMesh!");
-
-        //Create a serialized version of the solution vector
-//        NumericVector<Number> * serialized_solution = NumericVector<Number>::build(from_sys.comm()).release();
-//        serialized_solution->init(from_sys.n_dofs(), false, SERIAL);
-
-        // Need to pull down a full copy of this vector on every processor so we can get values in parallel
-//        from_sys.solution->localize(*serialized_solution);
+}
 
+void
+MultiAppCopyTransfer::transfer(FEProblem & to_problem, FEProblem & from_problem)
+{
+  // Populate the to/from variables needed to perform the transfer
+  _swapped = Moose::swapLibMeshComm(_multi_app->comm());
 
-        MeshBase * from_mesh = &from_problem.mesh().getMesh();
+  MooseVariable & to_var = to_problem.getVariable(0, _to_var_name);
+  MooseVariable & from_var = from_problem.getVariable(0, _from_var_name);
+  if (to_var.feType() != from_var.feType())
+    mooseError("The variables must be the same type (order and family).");
 
-        Moose::swapLibMeshComm(swapped);
+  _from_sys = &from_var.sys().system();
 
-        if (is_nodal)
-        {
-          MeshBase::const_node_iterator to_node_it = to_mesh->local_nodes_begin();
-          MeshBase::const_node_iterator to_node_end = to_mesh->local_nodes_end();
+  MeshBase * to_mesh = &to_problem.mesh().getMesh();
+  MeshBase * from_mesh = &from_problem.mesh().getMesh();
+  if ( (to_mesh->n_nodes() != from_mesh->n_nodes()) || (to_mesh->n_elem() != from_mesh->n_elem()))
+    mooseError("The meshes must be identical to utilize MultiAppCopyTransfer.");
 
-          for (; to_node_it != to_node_end; ++to_node_it)
-          {
-            Node * to_node = *to_node_it;
-            unsigned int to_node_id = to_node->id();
+  _from_sys_num = _from_sys->number();
+  _from_var_num = _from_sys->variable_number(from_var.name());
 
-            // The zero only works for LAGRANGE!
-            dof_id_type to_dof = to_node->dof_number(to_sys_num, to_var_num, 0);
+  System * to_sys = find_sys(to_problem.es(), _to_var_name);
+  _to_sys_num = to_sys->number();
+  _to_var_num = to_sys->variable_number(_to_var_name);
 
-            MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
+  _solution = to_problem.getNonlinearSystem().hasVariable(_to_var_name) ? &to_problem.getNonlinearSystem().solution() : &to_problem.getAuxiliarySystem().solution();
 
-            Node * from_node = from_mesh->node_ptr(to_node_id);
+  // Transfer node dofs
+  MeshBase::const_node_iterator node_it = to_mesh->local_nodes_begin();
+  MeshBase::const_node_iterator node_end = to_mesh->local_nodes_end();
+  for (; node_it != node_end; ++node_it)
+  {
+    _to_object = *node_it;
+    _from_object = from_mesh->node_ptr(_to_object->id());
+    transferDofObject();
```

Comment:
See - this is just not right... it's using "side effects".  The things to be operated on should be passed in.